source/constructs/lib/agent/unstructured-crawler/UnstructuredCrawler.py

```diff
@@ -0,0 +1,156 @@
+
+def postprocess_crawler_result(crawler_result, scan_depth):
+    """
+    Postprocesses the crawler result and returns the crawler result.
+    """
+    for key, value in crawler_result.items():
+        sample_size = scan_depth if scan_depth < len(value["sample_files"]) else len(value["sample_files"])
+        if len(value["sample_files"]) > sample_size:
+            value["sample_files"] = random.sample(value["sample_files"], sample_size)
+    return crawler_result
+
+def list_s3_objects(s3_client, bucket_name, scan_depth, include_file_extensions, exclude_file_extensions, prefix=''):
+    """
+    Lists the objects in the s3 bucket and returns the crawler result.
+    
+    Args:
+        s3_client: boto3 s3 client
+        bucket_name: name of the s3 bucket
+        scan_depth: number of files to be scanned
+        include_file_extensions: list of file extensions to be included in the scan
+        exclude_file_extensions: list of file extensions to be excluded in the scan
+        prefix: prefix of the s3 bucket
+    """
+    bucket_info = {}
+    detection_files, include_files, exclude_files = {}, {}, {}
+    paginator = s3_client.get_paginator('list_objects_v2')
+    pages = paginator.paginate(Bucket=bucket_name, Prefix=prefix, PaginationConfig={'PageSize': 1000})
+
+    supported_types = []
+    for file_extension in supported_text_types + suported_image_types:
+        supported_types.append(file_extension)
+
+    # iterate over pages
+    for page in pages:
+        # loop through objects in page
+        if 'Contents' in page:
+            page_detection_files, page_include_files, page_exclude_files = summarize_page(page, supported_types, include_file_extensions, exclude_file_extensions)
+            detection_files = update_dict_files(detection_files, page_detection_files)
+            include_files = update_dict_files(include_files, page_include_files)
+            exclude_files = update_dict_files(exclude_files, page_exclude_files)
+
+    detection_files = postprocess_crawler_result(detection_files, scan_depth)
+    include_files = postprocess_crawler_result(include_files, scan_depth)
+    exclude_files = postprocess_crawler_result(exclude_files, scan_depth)
+
+    # Get the current UTC time
+    now_utc = datetime.datetime.now(datetime.timezone.utc)
+    utc_string = now_utc.strftime('%Y-%m-%d %H:%M:%S.%f %Z')
+
+    bucket_info['bucket_name'] = bucket_name
+    bucket_info['last_updated_time'] = utc_string
+    bucket_info['detection_files'] = detection_files
+    bucket_info['include_files'] = include_files
+    bucket_info['exclude_files'] = exclude_files
+    bucket_info['include_file_extensions'] = include_file_extensions
+    bucket_info['exclude_file_extensions'] = exclude_file_extensions
+
+    return bucket_info
+
+def upload_bucket_info(s3_client, bucket_info, source_bucket_name, result_bucket_name, prefix=''):
+    # dump json format content to a file and save to s3
+    json_file_path = '/tmp/json_file.json'
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

It looks like you&#39;re creating or opening a file in global directory. Manipulating files from these directories can expose your application to race conditions. Also, an attacker can access, modify, corrupt, or delete files from these directories. To prevent this, use a dedicated sub-folder with controlled permission or secure-by-design APIs to create temporary files. For more information, see [CWE-379](https://cwe.mitre.org/data/definitions/379.html)

Reply:
fixed